### DISCUSSION TITLE
[QUESTION] How to monitor bandwidth usage for each volume

### DISCUSSION THREAD

**Discussion by badihi** [longhorn/longhorn#4196] [Q&A] [answered]
## Question
How can I monitor the network bandwidth that is being used by each volume to replicate and/or transfer the data to the mounted host? This will be so useful for planning locality and disk usage over the cluster.

## Environment

 - Longhorn version: 1.2.4
 - Kubernetes version: 1.21.5
 - Node config
   - OS type and version: Ubuntu 20
   - CPU per node: 8
   - Memory per node: 16
   - Disk type: SSD
   - Network bandwidth and latency between the nodes: 1Gbps
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): VMWare ESXI

**Comment by PhanLe1010** [comment]
Can setting up Prometheus to monitor the network of inter nodes help?
How about monitoring the network of instance-manger-xxx pods which host the engines and the replicas of volume?

What do you think?

**Comment by PhanLe1010** [comment]
The network bandwidth to move data from the engine of the volume to the replica of the volume can be calculated using the number of IOPs. However, Longhorn has not exposed this metrics for each volume yet

**Comment by badihi** [comment]
@PhanLe1010  I have some questions to make what you said clear:
- Do you confirm that the network throughput of the container `instance-manger-xxx` will be equal to throughput of all volumes on the node (the sum of replication and read/write operations)? If yes, this will be a good place to start.
- Do you mean that the exact network bandwidth can be derived from IOPS (if yes, please describe how), or it is just a measure to guess the amount of bandwidth? Is "IOPS done for replication" included in the metrics? Or is it just the r/w operations?

**Comment by PhanLe1010** [comment, marked as answer]
@badihi 
> Do you confirm that the network throughput of the container instance-manger-xxx will be equal to throughput of all volumes on the node (the sum of replication and read/write operations)? If yes, this will be a good place to start.

The sum of network throughput of the instance-manager-e-xxx pod and instance-manager-r-xxx pod is the upper bound of the throughput of all volumes on the node (the sum of replication and read/write operations)


> Do you mean that the exact network bandwidth can be derived from IOPS (if yes, please describe how), or it is just a measure to guess the amount of bandwidth? Is "IOPS done for replication" included in the metrics? Or is it just the r/w operations?

A Longhorn volume has an engine process and multiple replicas process. Data are moving between the engine (which lives inside instance-manager-e-xxx pod) and the the replicas (which lives inside instance-manager-r-xxx pods). Data are moved by blocks. Now, if you know the number of IOPs on this path, I think we can calculate the network throughput of the volume. However, Longhorn has not exposed this metrics yet.
